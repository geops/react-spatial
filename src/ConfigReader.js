import TileLayer from 'ol/layer/Tile';
import XYZ from 'ol/source/XYZ';
import VectorSource from 'ol/source/Vector';
import WMTSSource from 'ol/source/WMTS';
import WMTSTileGrid from 'ol/tilegrid/WMTS';
import GeoJSONFormat from 'ol/format/GeoJSON';
import Layer from './Layer';
import VectorLayer from './VectorLayer';

import projections from './Projections';

class ConfigReader {
  static createXYZLayer(item) {
    const conf = { ...item };
    delete conf.data;

    return new Layer({
      ...conf,
      olLayer: new TileLayer({
        zIndex: -1,
        source: new XYZ({
          url: item.data.url,
        }),
      }),
    });
  }

  static createVectorLayer(item) {
    const conf = { ...item };
    delete conf.data;

    return new VectorLayer({
      ...conf,
      source: new VectorSource({
        url: item.data.url,
        format: new GeoJSONFormat(),
      }),
      style: item.data.style,
    });
  }

  static createWMTSLayer(item) {
    const conf = { ...item };
    delete conf.data;

    return new Layer({
      ...conf,
      olLayer: new TileLayer({
        zIndex: -1,
        source: new WMTSSource({
          url: item.data.url,
          tileGrid: new WMTSTileGrid({
            extent: item.data.projectionExtent,
            resolutions: item.data.resolutions,
            matrixIds: item.data.resolutions.map((res, i) => `${i}`),
          }),
        }),
      }),
    });
  }

  static createEmptyLayer(item) {
    const conf = { ...item };
    delete conf.data;

    return new Layer({
      ...conf,
    });
  }

  static createLayer(item) {
    let layer;

    switch (item.data.type) {
      case 'xyz':
        layer = ConfigReader.createXYZLayer(item);
        break;
      case 'wmts':
        layer = ConfigReader.createWMTSLayer(item);
        break;
      case 'vectorLayer':
        layer = ConfigReader.createVectorLayer(item);
        break;
      default:
        layer = ConfigReader.createEmptyLayer(item);
    }

    return layer;
  }

  constructor(map, data, style = null) {
    this.map = map;
    this.data = data;
    this.style = style;

    this.defaults = {
      data: {},
      visible: false,
      isBaseLayer: false,
    };
  }

  loadLayerFromConfig(config) {
    // apply default values
    const item = { ...this.defaults, ...config };

    if (item.data && item.data.type === 'wmts') {
      const proj = projections[item.data.projection];
      if (proj) {
        item.data.projectionExtent = proj.projectionExtent;
        item.data.resolutions = proj.resolutions;
      }
    }

    const layer = ConfigReader.createLayer(item);
    layer.init(this.map);

    if (item.children) {
      item.children.forEach(childConf => {
        layer.addChild(this.loadLayerFromConfig(childConf));
      });
    }

    return layer;
  }

  initialize() {
    return this.data.map(val => this.loadLayerFromConfig(val));
  }
}

const readConfig = (map, data) => {
  const configReader = new ConfigReader(map, data);
  return configReader.initialize();
};

const getVisibleTopic = topicList => {
  let visibleTopic = null;
  topicList.forEach(topic => {
    if (topic.visible) {
      visibleTopic = topic;
    }
  });
  return visibleTopic.children;
};

export default { getVisibleTopic, readConfig };
