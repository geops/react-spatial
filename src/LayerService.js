import Observable from 'ol/Observable';

/**
 * A layer service class to handle layer adding, removing and visiblity.
 */
export default class LayerService {
  constructor(layers) {
    this.layers = layers;
    this.callbacks = {};
    this.callbacks = {};
    this.keys = [];
    this.listenChangeEvt();
  }

  getLayers() {
    return this.layers;
  }

  getLayersAsFlatArray(optLayers) {
    let layers = [];
    (optLayers || this.getLayers() || []).forEach(l => {
      layers.push(l);
      const children = l.getChildren();
      layers = layers.concat(this.getLayersAsFlatArray(children));
    });
    return layers;
  }

  getLayer(name) {
    return this.getLayersAsFlatArray().find(l => l.getName() === name);
  }

  getParent(child) {
    return this.getLayersAsFlatArray().find(
      l => !!l.getChildren().includes(child),
    );
  }

  getRadioGroupLayers(radioGroupName) {
    if (radioGroupName) {
      return this.getLayersAsFlatArray().filter(
        l => l.getRadioGroup() === radioGroupName,
      );
    }

    return null;
  }

  on(evt, callback) {
    this.callbacks[evt] = this.callbacks[evt] || [];
    this.callbacks[evt].push(callback);
  }

  listenChangeEvt() {
    this.getLayersAsFlatArray().forEach(layer => {
      this.keys.push(
        layer.on('change:visible', evt => {
          const visible = evt.target.getVisible();

          // Apply to siblings only if it's a radio group.
          if (
            !evt.stopPropagationSiblings &&
            layer.getRadioGroup() &&
            visible
          ) {
            const siblings = this.getRadioGroupLayers(
              layer.getRadioGroup(),
            ).filter(l => l !== layer);

            siblings.forEach(s => {
              if (
                visible &&
                s.getRadioGroup() &&
                evt.target.getRadioGroup() === s.getRadioGroup()
              ) {
                s.setVisible(false, false, true, true);
              }
            });
          }

          // Apply to children
          if (!evt.stopPropagationDown && layer.children) {
            layer.children.forEach(child => {
              child.setVisible(visible, false, true, false);
            });
          }

          // Apply to parent only if:
          //   - a child is visible
          //   - all children are hidden
          const parentLayer = this.getParent(layer);

          if (
            !evt.stopPropagationUp &&
            parentLayer &&
            (visible ||
              (!visible && !parentLayer.children.find(c => c.getVisible())))
          ) {
            parentLayer.setVisible(visible, true, false, false);
          }

          (this.callbacks['change:visible'] || []).forEach(cb => cb());
        }),
      );
    });
  }

  unlistenChangeEvt() {
    this.keys.forEach(key => {
      Observable.unByKey(key);
    });
  }
}
