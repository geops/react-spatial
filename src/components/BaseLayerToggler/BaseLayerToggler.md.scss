.tm-base-layer-example {
  position: relative;
  height: 300px;

  .tm-map {
    width: 100%;
    height: 100%;
    flex-grow: 1;
  }

  .tm-base-layer-toggler {
    position: absolute;
    bottom: 10px;
    left: 10px;
  }
}
