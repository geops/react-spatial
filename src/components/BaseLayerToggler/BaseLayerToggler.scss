.tm-base-layer-toggler {
  width: 120px;
  height: 120px;
  border: 2px solid lightgray;
  background-color: white;

  .tm-map {
    position: absolute;
    height: 100%;
    width: 100%;
  }

  .tm-base-layer-item {
    position: absolute;
    width: 100%;
    height: 100%;
  }

  .tm-footer {
    height: 21px;
    background-color: #fffa;
    display: flex;
    align-items: center;
    justify-content: space-between;
    padding: 0 2px;
    border: none;

    > div {
      display: flex;
      align-items: center;
    }

    .tm-base-layer-next,
    .tm-base-layer-previous {
      width: 18px;
      height: 18px;

      svg {
        width: 100%;
        height: 100%;
      }
    }

    .tm-base-layer-next {
      float: right;
    }

    .tm-base-layer-previous {
      float: left;
    }
  }
}
