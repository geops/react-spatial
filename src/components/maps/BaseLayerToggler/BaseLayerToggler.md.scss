.rs-base-layer-example {
  position: relative;

  .rs-base-layer-toggler {
    position: absolute;
    bottom: 10px;
    left: 10px;
  }
}
