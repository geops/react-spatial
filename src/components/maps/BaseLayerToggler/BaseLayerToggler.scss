.rs-base-layer-toggler {
  width: 120px;
  height: 120px;
  border: 2px solid lightgray;
  background-color: white;
  position: relative;

  .rs-base-layer-map,
  .rs-base-layer-image {
    position: absolute;
    height: 100%;
    width: 100%;
  }

  .rs-base-layer-toggle-button {
    width: 100%;
    height: 100%;
  }

  .rs-base-layer-footer {
    background-color: #fffa;
    position: absolute;
    bottom: 0;
    left: 0;
    right: 0;
    padding: 2px;
    height: 21px;
    z-index: 1;

    .rs-base-layer-next,
    .rs-base-layer-previous {
      width: 18px;
      height: 18px;

      svg {
        width: 100%;
        height: 100%;
      }
    }

    .rs-base-layer-next {
      float: right;
    }

    .rs-base-layer-previous {
      float: left;
    }
  }
}
