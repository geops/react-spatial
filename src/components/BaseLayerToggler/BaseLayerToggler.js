import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { FaArrowCircleLeft, FaArrowCircleRight } from 'react-icons/fa';
import OLMap from 'ol/Map';
import { unByKey } from 'ol/Observable';
import TileLayer from 'ol/layer/Tile';
import { containsExtent } from 'ol/extent';
import LayerService from '../../LayerService';
import Button from '../Button';
import Footer from '../Footer';
import BasicMap from '../BasicMap';
import MapboxLayer from '../../layers/MapboxLayer';

const propTypes = {
  /**
   * An ol map.
   */
  map: PropTypes.instanceOf(OLMap).isRequired,

  /**
   * Layers provider.
   */
  layerService: PropTypes.instanceOf(LayerService),

  /**
   * CSS class to apply on the container.
   */
  className: PropTypes.string,

  /**
   * CSS class to apply on each item.
   */
  classNameItem: PropTypes.string,

  /**
   * CSS class to apply to the previous button.
   */
  classNamePrevious: PropTypes.string,

  /**
   * CSS class to apply to the next button.
   */
  classNameNext: PropTypes.string,

  /**
   * Path to the directory which includes the fallback images
   */
  fallbackImgDir: PropTypes.string,

  /**
   * Outside of this valid extent the fallback image is loaded
   */
  validExtent: PropTypes.arrayOf(PropTypes.number),

  /**
   *  The tabIndex of the map.
   */
  mapTabIndex: PropTypes.number,

  /**
   * HTML tabIndex attribute of the button.
   */
  tabIndex: PropTypes.number,

  /**
   * title attribute of the baselayer button.
   */
  titleButton: PropTypes.string,

  /**
   * title attribute of the next baselayer button.
   */
  titleButtonNext: PropTypes.string,

  /**
   * title attribute of the previous baselayer button.
   */
  titleButtonPrevious: PropTypes.string,
};

const defaultProps = {
  layerService: undefined,
  className: 'tm-base-layer-toggler',
  classNameItem: 'tm-base-layer-item',
  classNamePrevious: 'tm-base-layer-previous',
  classNameNext: 'tm-base-layer-next',
  fallbackImgDir: '../../images/baselayer/',
  validExtent: [-Infinity, -Infinity, Infinity, Infinity],
  mapTabIndex: 0,
  tabIndex: 0,
  titleButton: 'Baselayer button',
  titleButtonNext: 'Next baselayer',
  titleButtonPrevious: 'Previous baselayer',
};

class BaseLayerToggler extends Component {
  constructor(props) {
    super(props);
    this.state = {
      layers: null,
      layerVisible: null,
      idx: 0,
      fallbackImg: null,
      fallbackImgOpacity: 0,
    };
    this.map = null;
    this.ref = React.createRef();
  }

  componentDidMount() {
    const { layerService, map } = this.props;

    if (layerService) {
      this.updateLayerService();
    }

    if (map) {
      this.updateMap();
    }
  }

  componentDidUpdate(prevProps, prevState) {
    const { layerService, map } = this.props;
    const { layerVisible, idx, layers } = this.state;

    if (layerService !== prevProps.layerService) {
      this.updateLayerService();
    }

    if (map !== prevProps.map) {
      this.updateMap();
    }

    if (layerVisible && !prevState.layerVisible) {
      this.next();
    } else if (layerVisible !== prevState.layerVisible) {
      // In case the visibility of the background Layer is change from another component.
      this.toggle(prevState.layerVisible);
    }

    if (this.map && idx !== prevState.idx) {
      this.map.getLayers().clear();

      const children = layers[idx].getChildren();
      const childLayers = children.length ? children : [layers[idx]];

      childLayers.forEach(layer => {
        if (layer instanceof MapboxLayer) {
          const ml = layer.clone();
          ml.init(this.map); // Including addLayer
          ml.setVisible(true);
        } else {
          this.map.addLayer(
            new TileLayer({
              source: layer.olLayer.getSource(),
            }),
          );
        }
        this.checkExtent();
      });
    }
  }

  componentWillUnmount() {
    unByKey([this.postRenderKey, this.moveEndKey]);
  }

  updateLayerService() {
    const { layerService } = this.props;

    if (!layerService) {
      return;
    }

    this.updateState();
    layerService.on('change:visible', () => this.updateState());
  }

  updateState() {
    const { layerService } = this.props;
    const layers = layerService.getBaseLayers() || [];
    let idx = layers.findIndex(l => l.getVisible());
    if (idx === -1 && layers.length > 1) {
      idx = 0;
      layers[idx].setVisible(true);
    }
    this.setState({
      layers,
      idx,
      layerVisible: layers.length > 1 ? layers[idx] : null,
    });
  }

  updateMap() {
    const { map } = this.props;
    if (!this.map) {
      this.map = new OLMap({ controls: [], interactions: [] });
    }
    this.map.setView(map.getView());
    unByKey([this.postRenderKey, this.moveEndKey]);
    this.postRenderKey = map.on('postrender', e => {
      this.map.getView().setZoom(e.target.getView().getZoom());
      if (this.ref && this.ref.current) {
        const elt = this.ref.current;
        const coord = map.getCoordinateFromPixel([
          elt.offsetLeft + elt.offsetWidth / 2,
          elt.offsetTop + elt.offsetHeight / 2,
        ]);
        this.map.getView().setCenter(coord);
      }
    });

    this.moveEndKey = map.on('moveend', () => {
      this.checkExtent();
    });
  }

  toggle(layer) {
    const { layers } = this.state;

    let index = 0;
    while (layer !== layers[index]) {
      index += 1;
    }

    this.setState({
      idx: index,
    });
  }

  previous() {
    const { layers, layerVisible, idx } = this.state;
    let previousIdx = idx - 1;

    while (!layers[previousIdx] || layerVisible === layers[previousIdx]) {
      previousIdx -= 1;
      if (previousIdx < 0) {
        previousIdx = layers.length - 1;
      }
    }

    this.setState({
      idx: previousIdx,
    });
  }

  next() {
    const { layers, layerVisible, idx } = this.state;
    let nextIdx = idx + 1;

    while (!layers[nextIdx] || layerVisible === layers[nextIdx]) {
      nextIdx += 1;
      if (nextIdx >= layers.length) {
        nextIdx = 0;
      }
    }

    this.setState({
      idx: nextIdx,
    });
  }

  /**
   * Check if the next layer is inside the global extent.
   * If not, try setting a global image.
   */
  checkExtent() {
    const { validExtent, map, fallbackImgDir } = this.props;
    const { idx, layers, fallbackImg } = this.state;
    const nextLayer = layers[idx];
    let opacity = 0;
    let img = fallbackImg;

    if (validExtent && this.ref && this.ref.current && nextLayer) {
      const elt = this.ref.current;
      const blCoord = map.getCoordinateFromPixel([
        elt.offsetLeft,
        elt.offsetTop + elt.offsetHeight,
      ]);
      const trCoord = map.getCoordinateFromPixel([
        elt.offsetLeft + elt.offsetWidth,
        elt.offsetTop,
      ]);

      if (!blCoord || !trCoord) {
        return;
      }

      if (!containsExtent(validExtent, [...blCoord, ...trCoord])) {
        opacity = 1;
        img = `${fallbackImgDir}${nextLayer.getKey()}.png`;
      }
    }

    this.setState({
      fallbackImg: img,
      fallbackImgOpacity: opacity,
    });
  }

  render() {
    const {
      className,
      classNameItem,
      classNamePrevious,
      classNameNext,
      mapTabIndex,
      tabIndex,
      titleButton,
      titleButtonNext,
      titleButtonPrevious,
    } = this.props;
    const { layers, idx, fallbackImg, fallbackImgOpacity } = this.state;

    let footer = null;

    if (!layers || layers.length < 2) {
      return null;
    }

    if (layers.length > 2) {
      footer = (
        <Footer>
          <Button
            className={classNamePrevious}
            onClick={() => this.previous()}
            title={titleButtonPrevious}
          >
            <FaArrowCircleLeft />
          </Button>
          <Button
            className={classNameNext}
            onClick={() => this.next()}
            title={titleButtonNext}
          >
            <FaArrowCircleRight />
          </Button>
        </Footer>
      );
    }

    const nextLayer = layers[idx];

    return (
      <div className={className} ref={this.ref}>
        <BasicMap map={this.map} tabIndex={mapTabIndex} />
        <img
          src={fallbackImg}
          alt={fallbackImg}
          style={{ opacity: fallbackImgOpacity }}
          className={classNameItem}
        />
        <Button
          className={classNameItem}
          tabIndex={tabIndex}
          title={titleButton}
          onClick={() => nextLayer.setVisible(true)}
        />

        {footer}
      </div>
    );
  }
}

BaseLayerToggler.propTypes = propTypes;
BaseLayerToggler.defaultProps = defaultProps;

export default BaseLayerToggler;
