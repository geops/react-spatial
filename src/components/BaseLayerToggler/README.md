#

This demonstrates the use of LayerTree.

```jsx
import React from 'react';
import BaseLayerToggler from 'react-spatial/components/BaseLayerToggler';
import BasicMap from 'react-spatial/components/BasicMap';
import LayerService from 'react-spatial/LayerService';
import OLMap from 'ol/Map';
import ConfigReader from 'react-spatial/ConfigReader';

class BaseLayerTogglerExample extends React.Component {
  constructor(props) {
    super(props);

    this.center = [1149722.7037660484, 6618091.313553318];
    this.map = new OLMap({ controls: [] });

    const layers = ConfigReader.readConfig(
      this.map,
      treeData,
    );

    this.layerService = new LayerService(layers);
  }

  render() {
    return (
      <div className="tm-base-layer-example">
        <BasicMap
          map={this.map}
          center={this.center}
          zoom={6}
        />
        <BaseLayerToggler
          map={this.map}
          layerService={this.layerService}
        />
      </div>
    );
  }
}

<BaseLayerTogglerExample />;
```
