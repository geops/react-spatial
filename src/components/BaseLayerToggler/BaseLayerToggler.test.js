/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'jest-canvas-mock';
import OLMap from 'ol/Map';
import renderer from 'react-test-renderer';
import BaseLayerToggler from './BaseLayerToggler';
import exampleData from '../../../data/ExampleData';
import data from '../../../data/TreeData';
import ConfigReader from '../../ConfigReader';
import LayerService from '../../LayerService';

configure({ adapter: new Adapter() });

const shallowComp = (newData, props) => {
  const map = new OLMap({});
  const layers = ConfigReader.readConfig(newData || data);
  const layerService = new LayerService(layers);
  return shallow(
    <BaseLayerToggler
      layerService={layerService}
      map={map}
      {...(props || {})}
    />,
  );
};
const mountComp = (newData, props) => {
  const map = new OLMap({});
  const layers = ConfigReader.readConfig(newData || data);
  const layerService = new LayerService(layers);
  return mount(
    <BaseLayerToggler
      layerService={layerService}
      map={map}
      {...(props || {})}
    />,
  );
};

const expectSnapshot = (newData, props) => {
  const map = new OLMap({});
  const layers = ConfigReader.readConfig(newData || data);
  const layerService = new LayerService(layers);
  const component = renderer.create(
    <BaseLayerToggler
      layerService={layerService}
      map={map}
      {...(props || {})}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
};

describe('BaseLayerToggler', () => {
  describe('matches snapshots', () => {
    test('using default properties.', () => {
      expectSnapshot();
    });
  });

  test('initialize correctly the state', () => {
    const wrapper = shallowComp();
    const comp = wrapper.instance();
    expect(comp.state.idx).toBe(1);
    expect(comp.state.layers.length).toBe(3);
    expect(comp.state.layerVisible).toBe(comp.state.layers[0]);
  });

  test('goes forward through all available layer except the current layer displayed on the map.', () => {
    const wrapper = shallowComp();
    const comp = wrapper.instance();
    const layerVisible = comp.state.layers[0];
    expect(comp.state.layers.length).toBe(3);
    expect(comp.state.layerVisible).toBe(layerVisible);
    expect(comp.state.idx).toBe(1);

    wrapper.find('.tm-base-layer-next').simulate('click');
    expect(comp.state.layerVisible).toBe(layerVisible);
    expect(comp.state.idx).toBe(2);

    // Layer at index 0 is displayed on the map so we must ignore it
    wrapper.find('.tm-base-layer-next').simulate('click');
    expect(comp.state.layerVisible).toBe(layerVisible);
    expect(comp.state.idx).toBe(1);
  });

  test('goes backward through all available layer except the current layer displayed on the map.', () => {
    const wrapper = shallowComp();
    const comp = wrapper.instance();
    const layerVisible = comp.state.layers[0];
    expect(comp.state.layers.length).toBe(3);
    expect(comp.state.layerVisible).toBe(layerVisible);
    expect(comp.state.idx).toBe(1);

    wrapper.find('.tm-base-layer-previous').simulate('click');
    expect(comp.state.layerVisible).toBe(layerVisible);
    expect(comp.state.idx).toBe(2);

    // Layer at index 0 is displayed on the map so we must ignore it
    wrapper.find('.tm-base-layer-previous').simulate('click');
    expect(comp.state.layerVisible).toBe(layerVisible);
    expect(comp.state.idx).toBe(1);
  });

  test('display on the map the layer clicked', () => {
    const wrapper = mountComp(data);
    const comp = wrapper.instance();
    expect(comp.state.layers.length).toBe(3);
    expect(comp.state.layers[0].getVisible()).toBe(true);
    wrapper
      .find('.tm-base-layer-item')
      .at(1)
      .simulate('click');
    wrapper.update();
    expect(comp.state.layers[0].getVisible()).toBe(false);
    expect(comp.state.layers[1].getVisible()).toBe(true);
    expect(comp.state.layerVisible).toBe(comp.state.layers[1]);
    expect(comp.state.idx).toBe(2);
  });

  test('hide baseLayerToggler if only one baselayer', () => {
    const wrapper = mountComp(exampleData);
    expect(wrapper.find('.tm-base-layer-item').exists()).toBe(false);
  });
});
