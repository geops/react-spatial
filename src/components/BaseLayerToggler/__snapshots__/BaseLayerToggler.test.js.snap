// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseLayerToggler matches snapshots using default properties. 1`] = `
<div
  className="tm-base-layer-toggler"
>
  <div
    className="tm-map"
    role="menu"
    tabIndex={0}
  />
  <div
    className="tm-base-layer-item"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  />
  <div
    className="tm-footer"
  >
    <div
      className="tm-base-layer-previous"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 504C119 504 8 393 8 256S119 8 256 8s248 111 248 248-111 248-248 248zm28.9-143.6L209.4 288H392c13.3 0 24-10.7 24-24v-16c0-13.3-10.7-24-24-24H209.4l75.5-72.4c9.7-9.3 9.9-24.8.4-34.3l-11-10.9c-9.4-9.4-24.6-9.4-33.9 0L107.7 239c-9.4 9.4-9.4 24.6 0 33.9l132.7 132.7c9.4 9.4 24.6 9.4 33.9 0l11-10.9c9.5-9.5 9.3-25-.4-34.3z"
        />
      </svg>
    </div>
    <div
      className="tm-base-layer-next"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zm-28.9 143.6l75.5 72.4H120c-13.3 0-24 10.7-24 24v16c0 13.3 10.7 24 24 24h182.6l-75.5 72.4c-9.7 9.3-9.9 24.8-.4 34.3l11 10.9c9.4 9.4 24.6 9.4 33.9 0L404.3 273c9.4-9.4 9.4-24.6 0-33.9L271.6 106.3c-9.4-9.4-24.6-9.4-33.9 0l-11 10.9c-9.5 9.6-9.3 25.1.4 34.4z"
        />
      </svg>
    </div>
  </div>
</div>
`;
